diff --git a/dix/window.c b/dix/window.c
index 1953f02..2f6bdcf 100644
--- a/dix/window.c
+++ b/dix/window.c
@@ -182,6 +182,8 @@ static Bool TileScreenSaver(ScreenPtr pScreen, int kind);
 
 #define SubStrSend(pWin,pParent) (StrSend(pWin) || SubSend(pParent))
 
+char* RootPPM = NULL;
+
 #ifdef COMPOSITE
 static const char *overlay_win_name = "<composite overlay>";
 #endif
@@ -400,6 +402,115 @@ SetWindowToDefaults(WindowPtr pWin)
 #endif
 }
 
+static int
+get_int(FILE *fp)
+{
+  int c = 0;
+
+  while ((c = getc(fp)) != EOF)
+    {
+      if (isspace(c))
+	continue;
+
+      if (c == '#')
+	while (c = getc(fp))
+	  if (c == EOF)
+	    return 0;
+	  else if (c == '\n')
+	    break;
+
+      if (isdigit(c))
+	{
+	  int val = c - '0';
+	  while ((c = getc(fp)) && isdigit(c))
+	    val = (val * 10) + (c - '0');
+	  return val;
+	}
+    }
+
+  return 0;
+}
+
+static unsigned char*
+ppm_load (const char* path, int depth, int *width, int *height)
+{
+  FILE *fp;
+  int   max, n = 0, w, h, i, j, bytes_per_line;
+  unsigned char *data, *res, h1, h2;
+
+  if (depth < 16 || depth > 32)
+    return NULL;
+
+  if (depth > 16)
+    depth = 32;
+
+  fp = fopen (path, "r");
+  if (fp == NULL)
+    return FALSE;
+
+  h1 = getc(fp);
+  h2 = getc(fp);
+
+  /* magic is 'P6' for raw ppm */
+  if (h1 != 'P' && h2 != '6')
+      goto fail;
+
+  w = get_int(fp);
+  h = get_int(fp);
+
+  if (w == 0 || h == 0)
+    goto fail;
+
+  max = get_int(fp);
+
+  if (max != 255)
+      goto fail;
+
+  bytes_per_line = ((w * depth + 31) >> 5) << 2;
+
+  res = data = malloc(bytes_per_line * h);
+
+  for (i=0; i<h; i++)
+    {
+      for (j=0; j<w; j++)
+	{
+	  unsigned char buf[3];
+	  fread(buf, 1, 3, fp);
+
+	  switch (depth)
+	    {
+	    case 24:
+	    case 32:
+	      *data     = buf[2];
+	      *(data+1) = buf[1];
+	      *(data+2) = buf[0];
+	      data += 4;
+	      break;
+	    case 16:
+	    default:
+	      *(unsigned short*)data
+		= ((buf[0] >> 3) << 11) | ((buf[1] >> 2) << 5) | (buf[2] >> 3);
+	      data += 2;
+	      break;
+	    }
+	}
+      data += (bytes_per_line - (w*(depth>>3)));
+    }
+
+  data = res;
+
+  *width  = w;
+  *height = h;
+
+  fclose(fp);
+
+  return res;
+
+ fail:
+  fclose(fp);
+  return NULL;
+}
+
 static void
 MakeRootTile(WindowPtr pWin)
 {
@@ -410,6 +521,36 @@ MakeRootTile(WindowPtr pWin)
     unsigned char *from, *to;
     int i, j;
 
+    if (RootPPM != NULL)
+      {
+	int            w, h;
+	unsigned char *data;
+
+	if ((data = ppm_load (RootPPM, pScreen->rootDepth, &w, &h)) != NULL)
+	  {
+	    pWin->background.pixmap
+	      = (*pScreen->CreatePixmap)(pScreen, w, h, pScreen->rootDepth, 0);
+
+	    pWin->backgroundState = BackgroundPixmap;
+	    pGC = GetScratchGC(pScreen->rootDepth, pScreen);
+	    if (!pWin->background.pixmap || !pGC)
+	      FatalError("could not create root tile");
+
+	    ValidateGC((DrawablePtr)pWin->background.pixmap, pGC);
+
+	    (*pGC->ops->PutImage)((DrawablePtr)pWin->background.pixmap,
+				  pGC,
+				  pScreen->rootDepth,
+				  0, 0, w, h, 0, ZPixmap, (char *)data);
+	    FreeScratchGC(pGC);
+
+	    free(data);
+	    return;
+	  }
+	else
+	  ErrorF("Unable to load root window image.");
+      }
+
     pWin->background.pixmap = (*pScreen->CreatePixmap)(pScreen, 4, 4,
 						    pScreen->rootDepth, 0);
 
@@ -626,6 +767,7 @@ RegisterRealChildHeadProc (RealChildHeadProc proc)
 }
 
 
+
 WindowPtr
 RealChildHead(WindowPtr pWin)
 {
diff --git a/hw/kdrive/src/kdrive.c b/hw/kdrive/src/kdrive.c
index 8dd039e..c62b7d0 100644
--- a/hw/kdrive/src/kdrive.c
+++ b/hw/kdrive/src/kdrive.c
@@ -60,6 +60,9 @@ KdDepths    kdDepths[] = {
     { 32, 32 }
 };
 
+int 
+ProcXFixesHideCursor (ClientPtr client) ;
+
 #define NUM_KD_DEPTHS (sizeof (kdDepths) / sizeof (kdDepths[0]))
 
 #define KD_DEFAULT_BUTTONS 5
@@ -91,6 +94,9 @@ static Bool         kdCaughtSignal = FALSE;
 
 KdOsFuncs	*kdOsFuncs;
 
+extern Bool CursorInitiallyHidden; /* See Xfixes cursor.c */
+extern char* RootPPM;		   /* dix/window.c */
+
 void
 KdDisableScreen (ScreenPtr pScreen)
 {
@@ -185,6 +191,7 @@ KdEnableScreen (ScreenPtr pScreen)
     SetRootClip (pScreen, TRUE);
     if (pScreenPriv->card->cfuncs->dpms)
 	(*pScreenPriv->card->cfuncs->dpms) (pScreen, pScreenPriv->dpmsState);
+
     return TRUE;
 }
 
@@ -454,6 +461,8 @@ KdUseMsg (void)
     ErrorF("-switchCmd       Command to execute on vt switch\n");
     ErrorF("-zap             Terminate server on Ctrl+Alt+Backspace\n");
     ErrorF("vtxx             Use virtual terminal xx instead of the next available\n");
+    ErrorF("-hide-cursor     Start with cursor hidden\n");
+    ErrorF("-root-ppm [path] Specify ppm file to use as root window background.\n");
 }
 
 int
@@ -517,6 +526,19 @@ KdProcessArgument (int argc, char **argv, int i)
 	kdSoftCursor = TRUE;
 	return 1;
     }
+    if (!strcmp (argv[i], "-hide-cursor"))
+    {
+      CursorInitiallyHidden = TRUE;
+      return 1;
+    }
+    if (!strcmp (argv[i], "-root-ppm"))
+    {
+      if ((i+1) < argc)
+	RootPPM =  argv[i+1];
+      else
+	UseMsg ();
+      return 2;
+    }
     if (!strcmp (argv[i], "-videoTest"))
     {
 	kdVideoTest = TRUE;
diff --git a/xfixes/cursor.c b/xfixes/cursor.c
index ecbed40..06febcd 100644
--- a/xfixes/cursor.c
+++ b/xfixes/cursor.c
@@ -60,6 +60,7 @@
 static RESTYPE		CursorClientType;
 static RESTYPE		CursorHideCountType;
 static RESTYPE		CursorWindowType;
+static Bool		CursorGloballyHidden;
 RESTYPE			PointerBarrierType;
 static CursorPtr	CursorCurrent[MAXDEVICES];
 
@@ -68,6 +69,8 @@ static DevPrivateKeyRec CursorScreenPrivateKeyRec;
 
 static void deleteCursorHideCountsForScreen (ScreenPtr pScreen);
 
+Bool CursorInitiallyHidden = FALSE;
+
 #define VERIFY_CURSOR(pCursor, cursor, client, access)			\
     do {								\
 	int err;							\
@@ -160,7 +163,7 @@ CursorDisplayCursor (DeviceIntPtr pDev,
     if (ConnectionInfo)
 	CursorVisible = EnableCursor;
 
-    if (cs->pCursorHideCounts != NULL || !CursorVisible) {
+    if (cs->pCursorHideCounts != NULL || !CursorVisible || CursorGloballyHidden) {
 	ret = (*pScreen->DisplayCursor) (pDev, pScreen, NullCursor);
     } else {
 	ret = (*pScreen->DisplayCursor) (pDev, pScreen, pCursor);
@@ -896,6 +899,12 @@ ProcXFixesHideCursor (ClientPtr client)
 	return ret;
     }
 
+    /* Is cursor set to be initially hidden ?, if so reset this
+     * flag as now visibility assumed under control of client.
+    */
+    if (CursorGloballyHidden)
+      CursorGloballyHidden = FALSE;
+
     /* 
      * Has client hidden the cursor before on this screen? 
      * If so, just increment the count. 
@@ -959,9 +968,19 @@ ProcXFixesShowCursor (ClientPtr client)
 	return rc;
     }
 
+    /* X was started with cursor hidden, therefore just reset our flag
+     * (returning to normal client control) and cause cursor to now be
+     * shown.
+    */
+    if (CursorGloballyHidden == TRUE)
+      {
+	CursorGloballyHidden = FALSE;
+	return (client->noClientException);
+      }
+
     /* 
      * Has client hidden the cursor on this screen?
-     * If not, generate an error.
+     * If so, generate an error.
      */
     pChc = findCursorHideCount(client, pWin->drawable.pScreen);
     if (pChc == NULL) {
@@ -1426,6 +1445,8 @@ XFixesCursorInit (void)
 {
     int	i;
 
+    CursorGloballyHidden = CursorInitiallyHidden;
+
     if (party_like_its_1989)
 	CursorVisible = EnableCursor;
     
