diff --git a/hw/kdrive/linux/keyboard.c b/hw/kdrive/linux/keyboard.c
index f5f9d9b..c207666 100644
--- a/hw/kdrive/linux/keyboard.c
+++ b/hw/kdrive/linux/keyboard.c
@@ -486,6 +486,8 @@ readKernelMapping(KdKeyboardInfo *ki)
     ki->minScanCode = minKeyCode;
     ki->maxScanCode = maxKeyCode;
 #endif
+    ki->minScanCode = 0;
+    ki->maxScanCode = NR_KEYS;
 }
 
 /*
@@ -707,11 +709,6 @@ LinuxKeyboardEnable (KdKeyboardInfo *ki)
     cfsetispeed(&nTty, 9600);
     cfsetospeed(&nTty, 9600);
     tcsetattr(fd, TCSANOW, &nTty);
-    /*
-     * Flush any pending keystrokes
-     */
-    while ((n = read (fd, buf, sizeof (buf))) > 0)
-	;
     KdRegisterFd (fd, LinuxKeyboardRead, ki);
     return Success;
 }
